### PR TITLE
fix more parsing underflows found while fuzzing

### PR DESCRIPTION
**Pull request by travisstaloch**
this patch allows these inputs to successfully parse and adds tests for them:
* `struct {}` or `union {}` - std.zig.Ast.firstToken()#L621

this is a draft because the added test doesn't work.  i don't know where to put the tests or how to change them so that they will lead to the same outcome as zig or zls.  here is my what i mean (copied from comment in test added by this pr).

> somehow testCanonical() causes these to be interpreted as
.container_field which safely leads to std.Ast.zig firstToken() L562.
but zls and zig interpret as .container_decl (i think) and both crash
on L621. this can be observed by placing either of these
contents in a file and running `zig test` on that file or opening that
file with zls.

### CODE REVIEW

**Comment by travisstaloch** [comment]
oops i thought a draft pr would prevent ci jobs from running. i'll see if i can cancel them.

**Comment by travisstaloch** [comment]
i don't think i can cancel them. feel free to cancel them if anyone sees this.

**Comment by travisstaloch** [comment]
closes https://github.com/zigtools/zls/issues/841

File: lib/std/zig/parser_test.zig [hunk 1]
```diff
@@ -228,6 +228,23 @@ test "zig fmt: top-level enum missing 'const name ='" {
     , &[_]Error{.expected_token});
 }
 
+test "zig fmt: top-level container missing 'const name ='" {
+    // somehow testCanonical() causes these to be interpreted as
```
Comment (r00ster91) [inline lib/std/zig/parser_test.zig:232]: I think this is kind of an awkward comment to add into the code itself. Also the line numbers in this comment will quickly be outdated with more changes to lib/std/zig/Ast.zig.

Comment (travisstaloch) [reply lib/std/zig/parser_test.zig:232]: yes i realize and plan on deleting it once i find a solution.

Comment (r00ster91) [reply lib/std/zig/parser_test.zig:232]: Whoops, sorry for barging in! I thought this was already solved because the tests pass for me.

Comment (travisstaloch) [reply lib/std/zig/parser_test.zig:232]: no problem. thanks for testing it out.  yeah its a bit confusing whats happening. i could have been more clear.  i thought those tests should fail without the change to Ast.zig. but they dont. you can observe it by removing the change to Ast.zig and running `zig test parser_test.zig`.

**Comment by r00ster91** [comment]
Oh I think the tests have to go somewhere else because running `zig fmt` on a file with `union {}` in it does not crash but it does crash with `zig run`, without your fix.

**Comment by r00ster91** [comment]
I think the tests could technically go into `test/behavior/error.zig` but I'm not sure it's the right place either.

**Comment by travisstaloch** [comment]
> Oh I think the tests have to go somewhere else

Agree. Thats what i'm hoping to get feedback on. Perhaps in test/compile_errors.zig ?

**Comment by r00ster91** [comment]
That sounds like the best place I can think of right now. Better than `test/behavior/error.zig` for sure.

**Comment by travisstaloch** [comment]
thanks. will try moving there.

**Comment by travisstaloch** [comment]
i added the following case to test/compile_errors.zig

```zig
    {
        const case = ctx.obj("top-level container missing 'const name ='", .{});

        case.addError(
            \\struct {}
            \\
        , &[_][]const u8{
            ":1:1: error: expected ',' after field",
        });
    }
```

i expected to see some kind of error. but this output says there were none:
```console
$ zig build test-cases -Dskip-non-native -Dskip-release -Dtest-filter="top-level container missing 'const name ='"
Test [1/87] test_0... compiler [1/1184] top-level container missing 'const name =' [1/1] update [2/3] parse/... 
Expected error not found:
================================================================================
:1:1: error: expected ',' after field
================================================================================
update_index=0
top-level container missing 'const name =' failed: WrongCompileErrors
1 tests failed
```

and also seeing basically the same output when adding this case:
```zig
    {
        const case = ctx.obj("top-level container missing 'const name ='", .{});
        case.backend = .stage2;

        case.addSourceFile("b.zig",
            \\struct {}
        );

        case.addError(
            \\pub export fn entry() void {
            \\    _ = @import("b.zig");
            \\}
        , &[_][]const u8{
            ":1:1: error: expected type expression, found 'struct'",
        });
    }
```

these both indicate that no error happened. 

not sure why no errors are happening as would be the case when running `zig run` or `zig test` on a file with these contents. could it be that a panic is happening and no error is being reported maybe? i'm not sure what to try next.

**Comment by Vexu** [comment]
There isn't supposed to be an error, that is a valid Zig file consisting of a file-level tuple-like struct with one field of type `struct {}`.

**Comment by travisstaloch** [comment]
> There isn't supposed to be an error, that is a valid Zig file consisting of a file-level tuple-like struct with one field of type `struct {}`.

hmm. maybe i'm making a mistake and using a wrong compiler version? otherwise, why doesn't this crash with underflow like with `zig run`? i can't seem to repro the crash in any tests and would like to see the crash happen to know that the fix is being tested.

**Comment by Vexu** [comment]
Your patch fixes the crash for me, maybe try a behavior test that checks the typeinfo of the file-level struct?

**Comment by travisstaloch** [comment]
> maybe try a behavior test that checks the typeinfo of the file-level struct?

i figured out why i was confused. i'm able to repro the crash now in test/compile_errors.zig but only for the input `union {}` and not `struct {}`.  yesterday i was getting confused because `struct {}` is being correctly accepted and wasn't checking `union {}`. 

@Vexu would you still like a behavior test for `struct {}` which verifies that it has one field named '0'?

**Comment by travisstaloch** [comment]
actually, i think this does belong in test/behavior and not test/compile_errors.zig because these are not expected to error. let me know if it belongs elsewhere.

**Comment by travisstaloch** [comment]
added skips for all non-llvm backends. let me know if any of these should be removed.

**Comment by Vexu** [comment]
Riscv, sparc, and x86 aren't tested in the CI and the C and wasm backends should be able to compile that test. If you have qemu and wasmtime you can test it locally using `-ofmt=c` and `-fno-LLVM -target <target>`.

**Comment by travisstaloch** [comment]
thanks! i can't tell whats causing the x86_64-linux-release failure. i can successfully run the following locally after making a release build (with `zig build -Drelease -Denable-llvm`) and after applying your last patch. 

```console
$ zig-out/bin/zig build test-behavior -fqemu -fwasmtime -Dstatic-llvm -Dtarget=native-native-musl
```

not sure if this has any value, but i can't run the following. when i try it + -Dstatic-llvm it immediately fails to compile. and as is it runs for maybe 5 minutes but then shows a ton of memory leaks related to plan9. 
```console
$ zig-out/bin/zig build test docs -fqemu -fwasmtime -Dtarget=native-native-musl
```

**Comment by Vexu** [comment]
> i can't tell whats causing the x86_64-linux-release failure

It's unrelated.



> then shows a ton of memory leaks related to plan9.

That would indicate that it's not actually using LLVM since that would disable the debug allocator which does the leak checking.

**Pull request merged**